Sources/StreamChat/Database/DTOs/MessageDTO.swift

```diff
@@ -331,15 +338,32 @@ extension NSManagedObjectContext: MessageDatabaseSession {
             payload.threadParticipants.map { try saveUser(payload: $0) }
         )
 
+        var channelDTO: ChannelDTO?
+
         if let channelPayload = payload.channel {
-            let channelDTO = try saveChannel(payload: channelPayload, query: nil)
-            dto.channel = channelDTO
+            channelDTO = try saveChannel(payload: channelPayload, query: nil)
         } else if let cid = cid {
-            let channelDTO = ChannelDTO.loadOrCreate(cid: cid, context: self)
-            dto.channel = channelDTO
+            channelDTO = ChannelDTO.loadOrCreate(cid: cid, context: self)
         } else {
             log.assertationFailure("Should never happen because either `cid` or `payload.channel` should be present.")
         }
+
+        if let channelDTO = channelDTO {
+            var messagesCreatedAt = channelDTO.messages.map(\.createdAt)
+            if dto.pinned {
+                // If message is pinned, should add it to pinnedMessages
+                channelDTO.pinnedMessages.insert(dto)
+            } else {
+                // If message is not pinned, or was unpinned, should be removed from pinnedMessages
+                channelDTO.pinnedMessages.remove(dto)
+                // If the current message dto is pinned, should not be included to oldestMessagetAt
+                messagesCreatedAt.append(dto.createdAt)
+            }
+            if let oldestMessageAt = messagesCreatedAt.min() {
+                channelDTO.oldestMessageAt = oldestMessageAt
+            }
+            dto.channel = channelDTO
```

Comment:
I'm confused by this. I think we shouldn't touch the `channelDTO.oldestMessageAt` here at all. That's the main point of it - to update it only if it comes from the regular channel query. Implemented like this, if a message that's not presented locally is updated, it will adjust `channelDTO.oldestMessageAt` and we have exactly the problem we're trying to work around. What we do in `ChannelDTO` is enough IMO.